libpromises/rlist.c

```diff
@@ -325,18 +325,50 @@ Rlist *RlistAppendRval(Rlist **start, Rval rval)
     return rp;
 }
 
-Rval RvalNew(const void *item, RvalType type)
+Rval RvalNewRewriter(const void *item, RvalType type, JsonElement *map)
 {
+    int max_size = 10*CF_BUFSIZE+1;
+    char buffer_from[max_size];
+    char buffer_to[max_size];
+
     switch (type)
     {
     case RVAL_TYPE_SCALAR:
-        return (Rval) { xstrdup(item), RVAL_TYPE_SCALAR };
+        if (NULL != map && JsonLength(map) > 0)
+        {
+            // BufferExpandStr and BufferExpandRegexes functions would
+            // be great instead of this CF_BUFSIZE ugliness.  ExpandScalar
+            Buffer *format = BufferNew();
+            strncpy(buffer_from, item, max_size);
+
+            JsonIterator iter = JsonIteratorInit(map);
+            const char *key;
+
+            while ((key = JsonIteratorNextKey(&iter)))
+            {
+                BufferPrintf(format, "$(%s)", key);
+                ReplaceStr(buffer_from, buffer_to, max_size, BufferData(format), JsonObjectGetAsString(map, key));
+                strncpy(buffer_from, buffer_to, max_size);
+                BufferPrintf(format, "${%s}", key);
+                ReplaceStr(buffer_from, buffer_to, max_size, BufferData(format), JsonObjectGetAsString(map, key));
+            }
```

Comment:
This might give unexpected results. Since you are only iterating over each key once, you might not catch recursively defined variables, and whether or not you do will depend on the order of the map. For example:

```
map = { a=b, b=c }

Iterations:
    key="a":
        "$($(a))" -> "$(b)"
    key="b":
        "$(b)" -> "c"
```

However, if we flip the order:

```
map = { b=c, a=b }

Iterations:
    key="b":
        "$($(a))" -> "$($(a))"
    key="a":
        "$($(a))" -> "$(b)"
```

which will not resolve into "c".

Such cases are surely handled elsewhere in the code. Being no evaluator expert, I don't know exactly where, but the function ``ExpandAndMapIteratorsFromScalar`` may be a place to start looking.

If it proves too difficult to reuse existing code, then this could be implemented differently: Instead of iterating over the keys, find all occurrences of ``$(...)`` in the string, and replace those after looking up the key in the map. When you are finished, repeat the whole process, since the replacements may have revealed new variables that need resolving. Only stop once no more replacements can be made, either because there are no ``$(...)`` references left, or because the remaining ones can't be found in the map.